### PR TITLE
SDR Classifier without densematrix

### PR DESCRIPTION
**Pull request by breznak**
as DenseMatrix is to be removed.
Uses map<UInt, map<UInt, Real>> instead.

- [x] implement WIP parts
- [x] fixes bug in upstream softmax
- [ ] ~measure SDR Classifier performance in Hotgym benchmark~ -> moved to #30 

Blocks more cleanup in #169

### CODE REVIEW

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 1]
```diff
@@ -150,11 +128,26 @@ void SDRClassifier::compute(UInt recordNum, const vector<UInt> &patternNZ,
...
       // update weights
       if (binary_search(steps_.begin(), steps_.end(), nSteps)) {
-        vector<Real64> error =
-            calculateError_(bucketIdxList, learnPatternNZ, nSteps);
+        const vector<Real64> error = calculateError_(bucketIdxList, learnPatternNZ, nSteps);
         Matrix &weights = weightMatrix_.at(nSteps);
         for (auto &bit : learnPatternNZ) {
-          weights.axby(bit, 1.0, alpha_, error);
+	  NTA_THROW << "SDRC implement axbi()";
+          //FIXME weights.axby(bit, 1.0, alpha_, error);
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO abxy()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 2]
```diff
@@ -192,12 +185,14 @@ void SDRClassifier::infer_(const vector<UInt> &patternNZ,
...
-        result->createVector(*nSteps, maxBucketIdx_ + 1, 0.0);
+    vector<Real64> *likelihoods = result->createVector(*nSteps, maxBucketIdx_ + 1, 0.0);
     for (auto &bit : patternNZ) {
-      const Matrix &weights = weightMatrix_.at(*nSteps);
-      add(likelihoods->begin(), likelihoods->end(), weights.begin(bit),
-          weights.begin(bit + 1));
+      Matrix &weights = weightMatrix_.at(*nSteps);
+      NTA_THROW << "SDRC implement add()";
+      //FIXME implement add from ArrayAlgo::add()
+      //add(likelihoods->begin(), likelihoods->end(),
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO add()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 3]
```diff
@@ -211,8 +206,11 @@ vector<Real64> SDRClassifier::calculateError_(const vector<UInt> &bucketIdxList,
 
   for (auto &bit : patternNZ) {
     const Matrix &weights = weightMatrix_.at(step);
-    add(likelihoods.begin(), likelihoods.end(), weights.begin(bit),
-        weights.begin(bit + 1));
+    NTA_THROW << "SDRC implement add()";
+    //FIXME add from ArrayAlgo
+    //add(likelihoods.begin(), likelihoods.end(),
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO add()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 4]
```diff
@@ -221,8 +219,10 @@ vector<Real64> SDRClassifier::calculateError_(const vector<UInt> &bucketIdxList,
   Real64 numCategories = (Real64)bucketIdxList.size();
   for (size_t i = 0; i < bucketIdxList.size(); i++)
     targetDistribution[bucketIdxList[i]] = 1.0 / numCategories;
-
-  axby(-1.0, likelihoods, 1.0, targetDistribution);
+  
+  NTA_THROW << "SDRC implement axbi()";
+  //FIXME axby from DenseMatrix
+  //axby(-1.0, likelihoods, 1.0, targetDistribution);
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO abxy()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 5]
```diff
@@ -232,8 +232,9 @@ void SDRClassifier::softmax_(vector<Real64>::iterator begin,
   for (auto itr = begin; itr != end; ++itr) {
     *itr -= *maxItr;
   }
-  range_exp(1.0, begin, end);
-  Real64 sum = accumulate(begin, end, 0.0);
+  NTA_THROW << "SDRC implement range_exp";
+  //FIXME implement: range_exp(1.0, begin, end);
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO range_exp()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 6]
```diff
@@ -248,6 +249,8 @@ void SDRClassifier::setVerbosity(UInt verbosity) { verbosity_ = verbosity; }
 Real64 SDRClassifier::getAlpha() const { return alpha_; }
 
 void SDRClassifier::save(ostream &outStream) const {
+  NTA_THROW << "SDRClassifier save() unimplemented!";
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO fix save()

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 7]
```diff
@@ -299,9 +302,12 @@ void SDRClassifier::save(ostream &outStream) const {
 
   // Write an ending marker.
   outStream << "~SDRClassifier" << endl;
+  */
 }
 
 void SDRClassifier::load(istream &inStream) {
+  NTA_THROW << "SDR Classifier load() unimplemented!";
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: TODO fix load()

File: src/nupic/algorithms/SDRClassifier.hpp [hunk 8]
```diff
@@ -45,9 +44,9 @@ typedef cla_classifier::ClassifierResult ClassifierResult;
 
 namespace sdr_classifier {
 
-const UInt sdrClassifierVersion = 1;
+const UInt sdrClassifierVersion = 2;
 
-typedef Dense<UInt, Real64> Matrix;
+typedef std::map<UInt, std::map<UInt, Real64>> Matrix; //Matrix[r][c] = 0.0d
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.hpp]: I'm looking for comments on implementation of this 2D matrix. It's a mapping from 2d [UInt][UInt] -> Real64; 
- used to be DenseMatrix
- now considering: 
  - `map<int, map<int, double>>`
  - Eigen: `Matrix<Real64, Dynamic, Dynamic>`
  - plain Real64[][] //we sometimes have to grow it
  - vector<vector<Real>>

I'm inclining to eigen::Matrix as we need here concept of rows, cols.

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 9]
```diff
@@ -209,33 +227,44 @@ vector<Real64> SDRClassifier::calculateError_(const vector<UInt> &bucketIdxList,
...
   return likelihoods;
 }
 
-void SDRClassifier::softmax_(vector<Real64>::iterator begin,
-                             vector<Real64>::iterator end) {
-  vector<Real64>::iterator maxItr = max_element(begin, end);
-  for (auto itr = begin; itr != end; ++itr) {
-    *itr -= *maxItr;
+
+void SDRClassifier::softmax_(const vector<Real64>::iterator begin,
```
Comment (breznak) [inline src/nupic/algorithms/SDRClassifier.cpp]: @htm-community/nupic-cpp I'd like your help reviewing this `softmax_` function. I think I found an error in upstream implementation. It is not a big deal (didn't even have a test to spot it, but let's make sure it's right)

- I'm reworking SDRClassifier, and its softmax_()
- wrote test SDRClassifierTest::testSoftmax to compare results (with results obtained from master) -> and we diff! 
  - added debug statements
  - backported test and debug to master, you can view it at temporary PR # 

- the "original" softmax_ version from master uses: 
  - range_exp()  from math/ArrayAlgo.hpp
    - Exp<double>() from math/Math.hpp 
  - accumulate() - I'm not sure if std::, or something from SparseMatrix::accumulate

Debug results from master: 
```
max=2.018
in=0
in=1
in=1.337
in=2.018
in=1.1
in=0.5
in=0.9
exp=0.132921
exp=0.361317
exp=0.506111
exp=1
exp=3.00417
exp=1.64872
exp=2.4596
sum=9.11284
div=0.0145861
div=0.0396492
div=0.0555382
div=0.109735
div=0.329663
div=0.180923
div=0.269905
[       OK ] SDRClassifierTest.testSoftmax (1 ms)
[----------] 1 test from SDRClassifierTest (1 ms total)
``` 

and from this PR:
```
max=2.018
in=0
exp=0.132921
in=1
exp=0.361317
in=1.337
exp=0.506111
in=2.018
exp=1
in=1.1
exp=0.399317
in=0.5
exp=0.21915
in=0.9
exp=0.326933
sum=2.94575
div=0.045123
div=0.122657
div=0.171811
div=0.339472
div=0.135557
div=0.0743953
div=0.110985
[  FAILED  ] SDRClassifierTest.testSoftmax

 1 FAILED TEST
```

You can see a strange behavior in the old/master implementation after "2.018 aka max elemen aka where exp=1": 
the old function gives exp > 1, while our does return < 1 (as it should). Because we normalized the values to the max element( x[i]-maxVal) , all values should be <= 1. 

TL;DR: 
 - not too critical (no test even before)
 - I think ours new is correct, just want to doublecheck

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 8 above)
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.hpp]: for now implemented using native `map<int, map<int, double>> map2d;` 
Eigen would be optional improvement for the future, if needed.

File: src/nupic/algorithms/SDRClassifier.cpp (hunk 9 above)
Comment (breznak) [reply src/nupic/algorithms/SDRClassifier.cpp]: I went ahead with our version

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 8 above)
Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/SDRClassifier.hpp]: I like this matrix type.  I prefer the standard library types because they're understood by everyone.

**Comment by breznak** [comment]
@ctrl-z-9000-times @dkeeney this is now ready, please review when you have time

File: src/nupic/algorithms/SDRClassifier.cpp [hunk 10]
```diff
@@ -318,20 +340,18 @@ void SDRClassifier::load(istream &inStream) {
   // Check the version.
   UInt version;
   inStream >> version;
-  NTA_CHECK(version <= 1);
+  NTA_CHECK(version == 2);
```
Comment (dkeeney) [inline src/nupic/algorithms/SDRClassifier.cpp]: agree, we don't need to be backward compatible with old versions of serialization.

File: src/nupic/algorithms/SDRClassifier.hpp (hunk 8 above)
Comment (dkeeney) [reply src/nupic/algorithms/SDRClassifier.hpp]: I agree std:: classes are always good but I don't have any experience with eigen so I cannot say which is better.

**Comment by dkeeney** [comment]
Now that I have a fix for Circle CI PR #176 , perhaps we should push this one more time after the fix is in the master so that we can check this PR in debug mode.
